File: lib/hawat.rb

require 'hawat/html'

class Hawat
  def initialize(log_path)
    @log_path = log_path
  end

  class NamedAggregate
    attr_reader :nodes

    def initialize(nodes)
      @nodes = nodes
    end
  
    def update(line)
      @nodes.each do |_, node|
        node.update(line)
      end
    end

    def merge(other)
      other.nodes.each do |name, other_node|
        if node = @nodes[name]
          node.merge(other_node)
        else
          @nodes[name] = other_node
        end
      end
    end

    def collect
      h = {}
      @nodes.each do |name, node|
        h[name] = node.collect
      end
      h
    end
  end

  class MethodAggregate
    attr_reader :methods

    def initialize(&terminal_generator)
      @terminal_generator = terminal_generator
      @methods = {}
    end

    def update(line)
      terminal = (@methods[line.http_method] ||= @terminal_generator[])
      terminal.update(line)
    end

    def merge(other)
      other.methods.each do |method, node|
        @methods[method] ||= @terminal_generator[]
        @methods[method].merge(node)
      end
    end

    def collect
      h = {}
      @methods.each do |name, terminal|
        h[name] = terminal.collect
      end
      h
    end
  end

  class TimeBucketerAggregate
    attr_reader :bucket_length, :buckets

    def initialize(bucket_length_in_seconds, &block)
      @bucket_length = bucket_length_in_seconds
      @buckets = []
      @aggregate_generator = block
    end

    class Bucket < Struct.new(:start, :finish, :aggregate)
      def merge(other)
        raise if start != other.start or finish != other.finish or aggregate.class != other.aggregate.class
        aggregate.merge(other.aggregate)
      end
    end

    def update(line)
      close = line.closed
      @beginning_of_day ||= Time.local(close.year, close.mon, close.day, 0, 0, 0)
      close_s = close.to_i
      close_s_today = close_s - @beginning_of_day.to_i
      bucket_i = close_s_today/@bucket_length
      @buckets[bucket_i] ||= Bucket.new(@beginning_of_day + bucket_i*@bucket_length,
                                        @beginning_of_day + bucket_i*@bucket_length,
                                        @aggregate_generator.call)
      @buckets[bucket_i].aggregate.update(line)
    end

    def merge(other)
      raise unless bucket_length == other.bucket_length
      other.buckets.each_with_index do |other_bucket, i|
        next unless other_bucket
        if bucket = @buckets[i]
          bucket.merge(other_bucket)
        else
          @buckets[i] = other_bucket
        end
      end
    end

    def collect
      h = {}
      @buckets.each do |b|
        if b
          h[b.start] = b.aggregate.collect
        end
      end
      h
    end
  end

  class StatisticsTerminal
    attr_reader :count, :total_duration, :max_duration, :min_duration, :statuses

    def initialize
      @count          = 0
      @total_duration = 0
      @max_duration   = 0
      @min_duration   = 100000000
      @statuses       = Hash.new {|h,k| h[k] = 0}
    end

    def update(line_data)
      @count += 1
      dur = line_data.total_time
      @total_duration += dur
      @max_duration = dur if dur > @max_duration
      @min_duration = dur if dur < @min_duration
      @statuses[line_data.status] += 1
    end

    def merge(other)
      @count          += other.count
      @total_duration += other.total_duration
      @statuses        = @statuses.merge(other.statuses)

      @max_duration = other.max_duration if other.max_duration > @max_duration
      @min_duration = other.min_duration if other.min_duration < @min_duration
    end

    def collect
      if @count == 0
        {"count" => 0}
      else
        {
          "count" => @count,
          "status" => @statuses,
          "duration" => {
            "min"  => @min_duration,
            "mean" => (@total_duration.to_f/@count).to_i,
            "max"  => @max_duration
          }
        }
      end
    end
  end

  class PathStatsAggregate
    class PathNode
      attr_reader :count, :children, :terminal

      def initialize(terminal_generator)
        @children = {}
        @count = 0
        @terminal_generator = terminal_generator
      end

      def add(path, line)
        @count += 1
        bits = path.split("/")
        if bits.length > 1
          name = bits[1]
          node = (@children[name] ||= PathNode.new(@terminal_generator))
          node.add("/" + bits[2..-1].join("/"), line)
        else
          terminal.update(line)
        end
      end

      def terminal
        @terminal ||= @terminal_generator.call
      end

      def reduce
        #new_children = {"*" => PathNode.new(@terminal_generator) }
        #@children.each do |slug, node|
          #if node.count < @count/19
            #new_children["*"].merge(node)
          #else
            #new_children[slug] = node
          #end
        #end
        #unless new_children["*"].count > 0
          #new_children.delete("*")
        #end
        #@children = new_children
        
        if @children.length > 15
          new_children = {"*" => PathNode.new(@terminal_generator) }
          @children.each do |slug, node|
            new_children["*"].merge(node)
          end
          @children = new_children
        end

        @children.each {|slug, node| node.reduce }
      end

      def merge(other)
        @count += other.count
        other.children.each do |slug, node|
          @children[slug] ||= node
          @children[slug].merge(node)
        end
        terminal.merge(other.terminal) if other.terminal
      end

      def inspect(indent=0, s="")
        @children.each do |slug, node|
          s << " "*indent + "/#{slug} (#{node.count})\n"
          node.inspect(indent + 2, s)
        end
        s
      end

      def collect(path, out)
        @children.each do |slug, node|
          node.collect(path + "/" + slug, out)
        end
        out[path] = terminal.collect
      end
    end

    def initialize(&terminal_generator)
      @node = PathNode.new(terminal_generator)
    end

    def update(line)
      @node.add(line.path.split("?").first, line)
    end

    def collect
      r = {}
      @node.children.each {|slug, node| node.reduce }
      @node.collect("", r)
      r
    end
  end

  class ConcurrencyTerminal
    def initialize
      @max_concurrency = 0
      @live_requests = []
    end

    def update(line)
      acc_s = line.accepted.to_r
      close_s = line.closed.to_r
      @live_requests.reject! {|l| l > close_s }
      @live_requests << acc_s
      conc = @live_requests.length
      if conc > @max_concurrency
        @max_concurrency = conc 
      end
    end

    def collect
      {
        "max" => @max_concurrency
      }
    end
  end

  class FrontendAggregate
    def initialize(&node_generator)
      @node_generator = node_generator
      @aggregates = {}
    end

    def update(line_data)
      node = (@aggregates[line_data.frontend] ||= @node_generator[])
      node.update(line_data)
    end

    def collect
      h = {}
      @aggregates.each do |frontend, node|
        h[frontend] = node.collect
      end
      h
    end
  end

  def default_aggregate
    NamedAggregate.new("stats" => StatisticsTerminal.new, "conc" => ConcurrencyTerminal.new)
  end

  def default_stats
    frontends = FrontendAggregate.new do
      NamedAggregate.new(
        "global" => default_aggregate,
        "global_series" => TimeBucketerAggregate.new(5) { default_aggregate },
        "paths" => PathStatsAggregate.new { 
          MethodAggregate.new {
            NamedAggregate.new(
              "all" => StatisticsTerminal.new,
              "series" => TimeBucketerAggregate.new(5) { StatisticsTerminal.new })}
        })
    end
    NamedAggregate.new(
      "global"        => default_aggregate,
      "global_series" => TimeBucketerAggregate.new(5) { default_aggregate },
      "frontends"     => frontends)
  end

  def each_line
    line_data = LineData.new
    File.open(@log_path) do |file|
      while line = file.gets
        if md = LINE_RE.match(line)
          line_data.md = md
          yield line_data
        else
          $stderr.puts "LINE DIDN'T MATCH: #{line}"
        end
      end
    end
  end

  def statistics
    stats = default_stats
    each_line do |line|
      stats.update(line)
    end
    stats.collect
  end

  # example 
  # Apr 28 00:00:15 dc1-live-lb1.srv.songkick.net haproxy[32647]: 10.32.75.139:53757 [28/Apr/2013:00:00:15.885] skweb skweb/dc1-live-frontend6_3000 0/0/9/9/20 200 5732 - - ---- 104/39/39/5/0 0/0 \"GET /favicon.ico HTTP/1.1\"\n"
  LINE_RE = /^(\w+ \d+ \d+:\d+:\d+) (\S+) haproxy\[(\d+)\]: ([\d\.]+):(\d+) \[(\d+)\/(\w+)\/(\d+):(\d+:\d+:\d+\.\d+)\] ([^ ]+) ([^ ]+)\/([^ ]+) (-?\d+)\/(-?\d+)\/(-?\d+)\/(-?\d+)\/(-?\d+) (\d+) (\d+) - - ([\w-]+) (-?\d+)\/(-?\d+)\/(-?\d+)\/(-?\d+)\/(-?\d+) (\d+)\/(\d+) "(\w+) ([^ ]+)/

  class LineData
    attr_accessor :md

    def haproxy_host; md[2]; end
    def haproxy_pid;  md[3]; end
    def haproxy_ip;  md[4]; end
    def haproxy_port;  md[5]; end

    def day;           md[6]; end
    def month; md[7]; end
    def year; md[8]; end
    def timestamp; md[9]; end

    def accepted
      hour, minute, sec, milli = *timestamp.split(/:|\./)
      Time.local(year.to_i, month, day.to_i, hour.to_i, minute.to_i, sec.to_i, milli.to_i*1000)
    end

    def closed
      dur_s = Rational(total_time, 1000)
      Time.at(accepted.to_r + dur_s)
    end

    def frontend; md[10]; end
    def backend; md[11]; end
    def server; md[12]; end

    def tq; md[13].to_i; end
    def tw; md[14].to_i; end
    def tc; md[15].to_i; end
    def tr; md[16].to_i; end
    def total_time; md[17].to_i; end

    def status; md[18]; end
    def bytes; md[19]; end
    def termination_state; md[20]; end

    def actconn; md[21]; end
    def feconn; md[22]; end
    def beconn; md[23]; end
    def srv_conn; md[24]; end
    def retries; md[25]; end

    def srv_queue; md[26]; end
    def backend_queue; md[27]; end

    def http_method; md[28]; end
    def path; md[29]; end
  end
end

File: hawat.go

package main

import (
  "bufio"
  "encoding/json"
  "flag"
  "fmt"
  "io"
  "math/rand"
  "os"
  "regexp"
  "runtime/pprof"
  "strconv"
  "strings"
  "time"
  "ring"

  "github.com/glenn-brown/golang-pkg-pcre/src/pkg/pcre"
)

// example 
// Apr 28 00:00:15 dc1-live-lb1.srv.songkick.net haproxy[32647]: 10.32.75.139:53757 [28/Apr/2013:00:00:15.885] skweb skweb/dc1-live-frontend6_3000 0/0/9/9/20 200 5732 - - ---- 104/39/39/5/0 0/0 \"GET /favicon.ico HTTP/1.1\"\n"
//LINE_RE = /^(\w+ \d+ \d+:\d+:\d+) (\S+) haproxy\[(\d+)\]: ([\d\.]+):(\d+) \[(\d+)\/(\w+)\/(\d+):(\d+:\d+:\d+\.\d+)\] ([^ ]+) ([^ ]+)\/([^ ]+) (-?\d+)\/(-?\d+)\/(-?\d+)\/(-?\d+)\/(-?\d+) (\d+) (\d+) - - ([\w-]+) (-?\d+)\/(-?\d+)\/(-?\d+)\/(-?\d+)\/(-?\d+) (\d+)\/(\d+) "(\w+) ([^ ]+)/

var LineRegexSource = "^(\\w+\\s+\\d+ \\d+:\\d+:\\d+) (\\S+) haproxy\\[(\\d+)\\]: ([\\d\\.]+):(\\d+) \\[(\\d+)\\/(\\w+)\\/(\\d+):(\\d+:\\d+:\\d+\\.\\d+)\\] ([^ ]+) ([^ ]+)\\/([^ ]+) (-?\\d+)\\/(-?\\d+)\\/(-?\\d+)\\/(-?\\d+)\\/(-?\\d+) (-?\\d+) (-?\\d+) - - ([\\w-]+) (-?\\d+)\\/(-?\\d+)\\/(-?\\d+)\\/(-?\\d+)\\/(-?\\d+) (\\d+)\\/(\\d+) \"(\\w+) ([^ ]+)"
var LineRegex = regexp.MustCompile(LineRegexSource)
var LineRegexPCRE = pcre.MustCompile(LineRegexSource, 0)

type Hawat struct {
  filePath string
  node     Node
}

type Line interface {
  haproxyHost() string
  haproxyPid() string
  haproxyIp() string
  haproxyPort() string
  day() string
  month() string
  year() string
  timestamp() string
  frontend() string
  backend() string
  server() string
  tq() string
  tw() string
  tc() string
  tr() string
  totalTimeDuration() time.Duration
  totalTime() int
  status() string
  bytes() string
  terminationState() string
  actconn() string
  feconn() string
  beconn() string
  srvConn() string
  retries() string
  srvQueue() string
  backendQueue() string
  httpMethod() string
  path() string
  Accepted() time.Time
  Closed() time.Time
}

type LineDataPCRE pcre.Matcher

func (md *LineDataPCRE) GroupString(i int) string {
  matcher := pcre.Matcher(*md)
  return matcher.GroupString(i)
}
func (md *LineDataPCRE) haproxyHost() string { return md.GroupString(2) }
func (md *LineDataPCRE) haproxyPid() string { return  md.GroupString(3) }
func (md *LineDataPCRE) haproxyIp() string { return  md.GroupString(4) }
func (md *LineDataPCRE) haproxyPort() string { return  md.GroupString(5) }
func (md *LineDataPCRE) day() string { return md.GroupString(6) }
func (md *LineDataPCRE) month() string { return md.GroupString(7) }
func (md *LineDataPCRE) year() string { return md.GroupString(8) }
func (md *LineDataPCRE) timestamp() string { return md.GroupString(9) }
func (md *LineDataPCRE) frontend() string { return md.GroupString(10) }
func (md *LineDataPCRE) backend() string { return md.GroupString(11) }
func (md *LineDataPCRE) server() string { return md.GroupString(12) }
func (md *LineDataPCRE) tq() string { return md.GroupString(13) }
func (md *LineDataPCRE) tw() string { return md.GroupString(14) }
func (md *LineDataPCRE) tc() string { return md.GroupString(15) }
func (md *LineDataPCRE) tr() string { return md.GroupString(16) }
func (md *LineDataPCRE) totalTimeDuration() time.Duration { d,_ := time.ParseDuration(md.GroupString(17) + "ms"); return d }
func (md *LineDataPCRE) totalTime() int { i, _ := strconv.Atoi(md.GroupString(17)); return i }
func (md *LineDataPCRE) status() string { return md.GroupString(18) }
func (md *LineDataPCRE) bytes() string { return md.GroupString(19) }
func (md *LineDataPCRE) terminationState() string { return md.GroupString(20) }
func (md *LineDataPCRE) actconn() string { return md.GroupString(21) }
func (md *LineDataPCRE) feconn() string { return md.GroupString(22) }
func (md *LineDataPCRE) beconn() string { return md.GroupString(23) }
func (md *LineDataPCRE) srvConn() string { return md.GroupString(24) }
func (md *LineDataPCRE) retries() string { return md.GroupString(25) }
func (md *LineDataPCRE) srvQueue() string { return md.GroupString(26) }
func (md *LineDataPCRE) backendQueue() string { return md.GroupString(27) }
func (md *LineDataPCRE) httpMethod() string { return md.GroupString(28) }
func (md *LineDataPCRE) path() string { return md.GroupString(29) }

func (md *LineDataPCRE) Accepted() time.Time {
  t,_ := time.Parse("02/Jan/2006:15:04:05.000", md.day() + "/" + md.month() + "/" + md.year() + ":" + md.timestamp())
  return t
}

func (md *LineDataPCRE) Closed() time.Time {
  a := md.Accepted()
  return a.Add(md.totalTimeDuration())
}

type LineData []string

func (md LineData) haproxyHost() string { return md[2] }
func (md LineData) haproxyPid() string { return  md[3] }
func (md LineData) haproxyIp() string { return  md[4] }
func (md LineData) haproxyPort() string { return  md[5] }
func (md LineData) day() string { return md[6] }
func (md LineData) month() string { return md[7] }
func (md LineData) year() string { return md[8] }
func (md LineData) timestamp() string { return md[9] }
func (md LineData) frontend() string { return md[10] }
func (md LineData) backend() string { return md[11] }
func (md LineData) server() string { return md[12] }
func (md LineData) tq() string { return md[13] }
func (md LineData) tw() string { return md[14] }
func (md LineData) tc() string { return md[15] }
func (md LineData) tr() string { return md[16] }
func (md LineData) totalTimeDuration() time.Duration { d,_ := time.ParseDuration(md[17] + "ms"); return d }
func (md LineData) totalTime() int { i, _ := strconv.Atoi(md[17]); return i }
func (md LineData) status() string { return md[18] }
func (md LineData) bytes() string { return md[19] }
func (md LineData) terminationState() string { return md[20] }
func (md LineData) actconn() string { return md[21] }
func (md LineData) feconn() string { return md[22] }
func (md LineData) beconn() string { return md[23] }
func (md LineData) srvConn() string { return md[24] }
func (md LineData) retries() string { return md[25] }
func (md LineData) srvQueue() string { return md[26] }
func (md LineData) backendQueue() string { return md[27] }
func (md LineData) httpMethod() string { return md[28] }
func (md LineData) path() string { return md[29] }

func (md LineData) Accepted() time.Time {
  t,_ := time.Parse("02/Jan/2006:15:04:05.000", md.day() + "/" + md.month() + "/" + md.year() + ":" + md.timestamp())
  return t
}

func (md LineData) Closed() time.Time {
  a := md.Accepted()
  return a.Add(md.totalTimeDuration())
}

type Node interface {
  Update(Line)
  Collect()         map[string]interface{}
  Merge(Node)
  Children()        map[string]Node
}

type NamedAggregate struct {
  _children map[string]Node
}

type SplitterAggregate struct {
  splitter       func(Line)string
  _children       map[string]Node
  childGenerator func()Node
}

type FrontendAggregate struct {
  _children       map[string]Node
  childGenerator func()Node
}

type PathAggregate struct {
  count             int
  terminal          Node
  _children         map[string]Node
  terminalGenerator func()Node
  depth             int
}

type TimeBucket struct {
  start   time.Time
  end     time.Time
  child   Node
}

type TimeBucketerAggregate struct {
  bucketLength   int
  buckets        map[int]TimeBucket
  childGenerator func()Node
  startOfDay     time.Time
}

type StatisticsTerminal struct {
  count           int
  totalDuration   int
  maxDuration     int
  statusCounts    map[string]int
}

type ConcurrencyTerminal struct {
  maxConcurrency   int
  liveRequests     *ring.RingInt64
  //liveRequests     map[time.Time]int
}

// NamedAggregate

func newNamedAggregate(_children map[string]Node) *NamedAggregate {
  return &NamedAggregate{_children}
}

func (n *NamedAggregate) Children() map[string]Node { return n._children }

func (n *NamedAggregate) Update(l Line) {
  for _, child := range n._children {
    child.Update(l)
  }
}

func (n *NamedAggregate) Collect() map[string]interface{} {
  result := make(map[string]interface{})
  for name, child := range n._children {
    result[name] = child.Collect()
  }
  return result
}

func (n *NamedAggregate) Merge(other Node) {
  for name, otherChild := range(other.Children()) {
    thisChild, ok := n._children[name]
    if ok {
      thisChild.Merge(otherChild)
    } else {
      n._children[name] = otherChild
    }
  }
}

// MethodAggregate

func newMethodAggregate(childGenerator func()Node) *SplitterAggregate {
  return newSplitterAggregate(func(ld Line)string { return ld.httpMethod() }, childGenerator)
}

// FrontendAggregate

func newFrontendAggregate(childGenerator func()Node) *SplitterAggregate {
  return newSplitterAggregate(func(ld Line)string { return ld.frontend() }, childGenerator)
}

// SplitterAggregate

func newSplitterAggregate(splitter func(Line)string, childGenerator func()Node) *SplitterAggregate {
  return &SplitterAggregate{splitter, make(map[string]Node), childGenerator}
}

func (n *SplitterAggregate) Children() map[string]Node { return n._children }

func (n *SplitterAggregate) Update(l Line) {
  child, ok := n._children[n.splitter(l)]
  if !ok {
    child = n.childGenerator()
    n._children[n.splitter(l)] = child
  }
  child.Update(l)
}

func (n *SplitterAggregate) Collect() map[string]interface{} {
  result := make(map[string]interface{})
  for method, child := range n._children {
    result[method] = child.Collect()
  }
  return result
}

func (n *SplitterAggregate) Merge(other Node) {
  for method, otherChild := range(other.Children()) {
    thisChild, ok := n._children[method]
    if ok {
      thisChild.Merge(otherChild)
    } else {
      n._children[method] = otherChild
    }
  }
}

// PathAggregate

func newPathAggregate(terminalGenerator func()Node) *PathAggregate {
  return newPathAggregateWithDepth(terminalGenerator, 0)
}

func newPathAggregateWithDepth(terminalGenerator func()Node, depth int) *PathAggregate {
  return &PathAggregate{0, terminalGenerator(), make(map[string]Node), terminalGenerator, depth}
}

func (n *PathAggregate) Children() map[string]Node        { return n._children }

func (n *PathAggregate) Update(l Line) {
  path := l.path()
  bits := strings.Split(path, "/")[1:]
  n.Update1(bits, l)
}

func (n *PathAggregate) Update1(bits []string, l Line) {
  n.count++
  if len(bits) > 1 {
    name := bits[0]
    node, ok := n._children["*"]
    if !ok {
      node, ok = n._children[name]
      if rand.Int31n(100) < 10 {
        n.Reduce()
      }
    }

    var child *PathAggregate
    if !ok {
      child = newPathAggregateWithDepth(n.terminalGenerator, n.depth + 1)
      n._children[name] = child
    } else {
      child = node.(*PathAggregate)
    }
    child.Update1(bits[1:], l)
  } else {
    n.terminal.Update(l)
  }
}

func (n *PathAggregate) Collect() map[string]interface{} {
  for _, child := range(n._children) {
    child.(*PathAggregate).Reduce()
  }
  result := map[string]interface{} {}
  n.Collect1("", result)
  return result
}

func (n *PathAggregate) Collect1(path string, result map[string]interface{}) {
  for slug, child := range(n._children) {
    child.(*PathAggregate).Collect1(path + "/" + slug, result)
  }
  result[path] = n.terminal.Collect()
}

func (n *PathAggregate) Reduce() {
  if n.depth == 0 {
    return
  }
  if _, ok := n._children["*"]; !ok {
    if len(n._children) > 15 {
      newChildren := map[string]Node { "*": newPathAggregate(n.terminalGenerator) }
      for _, child := range(n._children) {
        newChildren["*"].Merge(child)
      }
      n._children = newChildren
    }
    for _, child := range(n._children) {
      child.(*PathAggregate).Reduce()
    }
  }
}

func (n *PathAggregate) Merge(otherNode Node) {
  other := otherNode.(*PathAggregate)
  n.count += other.count
  for slug, otherChild := range(other.Children()) {
    thisChild, ok := n._children[slug]
    if !ok {
      thisChild = otherChild.(*PathAggregate)
      n._children[slug] = thisChild
    }
    thisChild.Merge(otherChild)
  }
  if other.terminal != nil {
    n.terminal.Merge(other.terminal)
  }
}

// TimeBucketerAggregate

func newTimeBucketerAggregate(bucketLength int, childGenerator func()Node ) *TimeBucketerAggregate {
  return &TimeBucketerAggregate{bucketLength, make(map[int]TimeBucket), childGenerator, time.Date(1970, 1, 1, 0, 0, 0, 0, time.UTC)}
}

func (n *TimeBucketerAggregate) Children() map[string]Node  { return nil }

func (n *TimeBucketerAggregate) Update(l Line) {
  closed := l.Closed()
  startOfDay := n.startOfDay
  if startOfDay.Year() == 1970 {
    startOfDay = time.Date(closed.Year(), closed.Month(), closed.Day(), 0, 0, 0, 0, startOfDay.Location())
    n.startOfDay = startOfDay
  }

  closedSecsToday := closed.Unix() - startOfDay.Unix()
  bucketIx := int(int(closedSecsToday)/n.bucketLength)
  bucket, ok := n.buckets[bucketIx]
  if !ok {
    start, _ := time.ParseDuration(strconv.Itoa(bucketIx*n.bucketLength) + "s")
    end, _ := time.ParseDuration(strconv.Itoa((bucketIx + 1)*n.bucketLength) + "s")
    bucket = TimeBucket{startOfDay.Add(start),
                        startOfDay.Add(end),
                        n.childGenerator()}
    n.buckets[bucketIx] = bucket
  }
  bucket.child.Update(l)
}

func (n *TimeBucketerAggregate) Collect() map[string]interface{} {
  result := map[string]interface{} {}
  for _, bucket := range(n.buckets) {
    result[bucket.start.Format(time.RubyDate)] = bucket.child.Collect()
  }
  return result
}

func (n *TimeBucketerAggregate) Merge(otherNode Node) {
  other := otherNode.(*TimeBucketerAggregate)
  for bucketIx, otherBucket := range other.buckets {
    if thisBucket, ok := n.buckets[bucketIx]; ok {
      thisBucket.Merge(&otherBucket)
    } else {
      n.buckets[bucketIx] = otherBucket
    }
  }
}

func (b TimeBucket) Merge(other *TimeBucket) {
  if !b.start.Equal(other.start) || !b.end.Equal(other.end) {
    panic("bucket start or end aren't equal")
  }
  b.child.Merge(other.child)
}

// StatisticsTerminal

func newStatisticsTerminal() *StatisticsTerminal {
  return &StatisticsTerminal{0, 0, 0, make(map[string]int)}
}

func (n *StatisticsTerminal) Children() map[string]Node        { return nil }

func (n *StatisticsTerminal) Update(l Line) {
  n.count++
  dur := l.totalTime()
  n.totalDuration += dur
  if dur > n.maxDuration {
    n.maxDuration = dur
  }
  n.statusCounts[l.status()] += 1
}

func (n *StatisticsTerminal) Collect() map[string]interface{} {
  if n.count == 0 {
    return map[string]interface{} {
      "count": 0,
    }
  } else {
    return map[string]interface{} {
      "count": n.count,
      "status": n.statusCounts,
      "duration": map[string]interface{} {
        "mean": n.totalDuration/n.count,
        "max": n.maxDuration,
      },
    }
  }
}

func (n *StatisticsTerminal) Merge(otherNode Node) {
  other := otherNode.(*StatisticsTerminal)
  n.count += other.count
  n.totalDuration += other.totalDuration
  if other.maxDuration > n.maxDuration {
    n.maxDuration = other.maxDuration
  }
  for status, otherCount := range(other.statusCounts) {
    thisCount, ok := n.statusCounts[status]
    if !ok {
      thisCount = 0
    }
    thisCount += otherCount
    n.statusCounts[status] = thisCount
  }
}

// ConcurrencyTerminal

func newConcurrencyTerminal() *ConcurrencyTerminal {
  return &ConcurrencyTerminal{0, ring.NewInt64(10000)}
  //return &ConcurrencyTerminal{0, make(map[time.Time]int)}
}

func (n *ConcurrencyTerminal) Children() map[string]Node { return nil }

func (n *ConcurrencyTerminal) Update(l Line) {
  accepted := l.Accepted()
  closed   := l.Closed()

  n.liveRequests.DeleteLessEq(-1*closed.UnixNano())
  n.liveRequests.Add(-1*accepted.UnixNano())
  if n.liveRequests.Length() > n.maxConcurrency {
    n.maxConcurrency = n.liveRequests.Length()
  }

  //conc := 0
  //for t, count := range(n.liveRequests) {
    //if t.After(closed) {
      //delete(n.liveRequests, t)
    //} else {
      //conc += count
    //}
  //}
  //n.liveRequests[accepted]++
  //if conc > n.maxConcurrency {
    //n.maxConcurrency = conc
  //}
}

func (n *ConcurrencyTerminal) Collect() map[string]interface{} {
  return map[string]interface{} {"max": n.maxConcurrency}
}

func (n *ConcurrencyTerminal) Merge(otherNode Node) {
  other := otherNode.(*ConcurrencyTerminal)
  if other.maxConcurrency > n.maxConcurrency {
    n.maxConcurrency = other.maxConcurrency
  }
}
func usage() {
  fmt.Printf("usage: hawat LOG_FILE\n")
}

func newHawat(filePath string) *Hawat {
  h := &Hawat{filePath, nil}
  return h
}

func newDefaultTerminal() *NamedAggregate {
  return newNamedAggregate(map[string]Node {
            "stats": newStatisticsTerminal(),
            "conc": newConcurrencyTerminal(),
         })
}

const TimeBucketLength = 300

func (h *Hawat) process() {
  file, _ := os.Open(h.filePath)
  reader := bufio.NewReader(file)

  frontends := newFrontendAggregate(func()Node {
    return newNamedAggregate(map[string]Node {
      "global": newNamedAggregate(map[string]Node {
        "all": newDefaultTerminal(),
        "series": newTimeBucketerAggregate(TimeBucketLength, func()Node { return newDefaultTerminal() })}),
      "paths": newPathAggregate(func()Node {
        return newMethodAggregate(func() Node {
          return newNamedAggregate(map[string]Node {
            "all": newDefaultTerminal(),
            "series": newTimeBucketerAggregate(TimeBucketLength, func()Node { return newStatisticsTerminal()})})})})})})

  node := newNamedAggregate(map[string]Node {
                "global": newNamedAggregate(map[string]Node {
                  "all": newDefaultTerminal(),
                  "series": newTimeBucketerAggregate(TimeBucketLength, func()Node { return newDefaultTerminal() })}),
                "frontends": frontends})
  i := 0

  for {
    i++
    line, err := reader.ReadString('\n')
    if err == io.EOF {
      break
    } else {
      processLine(line, node)
    }
  }
  b, _ := json.MarshalIndent(node.Collect(), "", "  ")
  fmt.Print(string(b) + "\n")
}

func processLine(line string, node Node) {
  defer func() {
    if x := recover(); x != nil {
      //fmt.Println("panic")
      //fmt.Println(line)
    }
  }()
  processLinePCRE(line, node)
}

func processLineNative(line string, node Node) {
  match := LineRegex.FindStringSubmatch(line)
  if match != nil {
    ld := LineData(match)
    ld.day()
    //node.Update(ld)
  } else {
    //fmt.Println("didn't match:")
    //fmt.Println(line)
  }
}

func processLinePCRE(line string, node Node) {
  if matcher := LineRegexPCRE.MatcherString(line, 0); matcher != nil {
    ld := LineDataPCRE(*matcher)
    node.Update(&ld)
  }
}

func main() {
  var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")
  var memprofile = flag.String("memprofile", "", "write memory profile to this file")

  flag.Parse()
  if *cpuprofile != "" {
    fmt.Println("profiling")
    f, err := os.Create(*cpuprofile)
    if err != nil {
      fmt.Println(err)
    }
    pprof.StartCPUProfile(f)
    defer pprof.StopCPUProfile()
  }

  if len(flag.Args()) < 1 {
    usage()
  } else {
    path := flag.Args()[0]
    h := newHawat(path)

    if *memprofile != "" {
      f, err := os.Create(*memprofile)
      if err != nil {
        fmt.Println(err)
      }
      h.process()
      pprof.WriteHeapProfile(f)
      f.Close()
      return
    } else {
      h.process()
    }

  }
}




